### PR TITLE
[NNAPI EP] Add EP option to disable CPU

### PR DESCRIPTION
**Pull request by gwang-msft**
**Description**: [NNAPI EP] Add EP option to disable CPU

**Motivation and Context**
- Add option for NNAPI EP to disable the execution on CPU, when it is not possible, the NNAPI EP compilation on the particular fused node will fail
- This is related to this change #6580

### CODE REVIEW

File: onnxruntime/core/providers/nnapi/nnapi_builtin/builders/model_builder.cc [hunk 1]
```diff
@@ -515,7 +522,34 @@ Status ModelBuilder::Compile(std::unique_ptr<Model>& model) {
       nnapi_->ANeuralNetworksModel_finish(nnapi_model_->model_),
       "on model finish");
 
+  // We have a list of target devices, try to see if the model can be run entirely
+  // using the list of target devices
+  bool use_create_for_devices = false;
   if (!nnapi_target_devices_.empty()) {
+    bool supported_ops[num_nnapi_ops_];
```
Comment (edgchen1) [inline onnxruntime/core/providers/nnapi/nnapi_builtin/builders/model_builder.cc:529]: i think that the array needs to have a constant value as the dimension. interestingly, it seems only MSVC fails to compile something like this. https://godbolt.org/z/G96c3Y
maybe use std::unique_ptr<bool[]> instead?

Comment (gwang-msft) [reply onnxruntime/core/providers/nnapi/nnapi_builtin/builders/model_builder.cc:529]: I didn't know that, changed to use std::unique_ptr<bool[]>

Comment (edgchen1) [reply onnxruntime/core/providers/nnapi/nnapi_builtin/builders/model_builder.cc]: cool, would also suggest using onnxruntime::make_unique<bool[]>(num_nnapi_ops_)

File: include/onnxruntime/core/providers/nnapi/nnapi_provider_factory.h [hunk 2]
```diff
@@ -19,9 +19,15 @@ enum NNAPIFlags {
   // Please note for now, NNAPI perform worse using NCHW compare to using NHWC
   NNAPI_FLAG_USE_NCHW = 0x002,
 
+  // Disable using NNAPI CPU implementation (nnapi-reference) for execution
+  // Please note, by enabling this option, it may cause the model cannot be executed using
```
Comment (skottmckay) [inline include/onnxruntime/core/providers/nnapi/nnapi_provider_factory.h:23]: Possibly not clear enough given there's NNAPI, the NNAPI EP and our CPU implementations. Is '(nnapi-reference)' meant to be a link to some NNAPI doco?

Maybe something a bit more verbose/explicit like
>Prevent NNAPI from using CPU devices. If NNAPI would use a CPU device to execute a node, and this flag is set, use the ORT kernel instead.

Also need a comment explaining why someone would set this.

File: onnxruntime/core/providers/nnapi/nnapi_builtin/builders/model_builder.cc [hunk 3]
```diff
@@ -515,7 +522,37 @@ Status ModelBuilder::Compile(std::unique_ptr<Model>& model) {
...
+    auto* supported_ops = supported_ops_holder.get();
+    RETURN_STATUS_ON_ERROR_WITH_NOTE(
+        nnapi_->ANeuralNetworksModel_getSupportedOperationsForDevices(
+            nnapi_model_->model_, nnapi_target_devices_.data(),
+            nnapi_target_devices_.size(), supported_ops),
+        "on getSupportedOperationsForDevices");
+
+    auto* it = std::find(supported_ops, supported_ops + num_nnapi_ops_, false);
+    if (it != supported_ops + num_nnapi_ops_) {
+      // There are some ops not supported by the list of the target devices
```
Comment (skottmckay) [inline onnxruntime/core/providers/nnapi/nnapi_builtin/builders/model_builder.cc:541]: nit: std::any_of may be slightly clearer

Comment (gwang-msft) [reply onnxruntime/core/providers/nnapi/nnapi_builtin/builders/model_builder.cc:541]: Change to std::all_of

**Review by skottmckay — APPROVED** [review]

**Pull request merged**